ws-security-dom/src/main/java/org/apache/wss4j/dom/processor/EncryptedKeyProcessor.java

```diff
@@ -276,27 +275,42 @@ private CertificateResult getPublicKey(Element keyValueElement, RequestData data
             X509Certificate[] certs = getCertificatesFromX509Data(keyValueElement, data);
             builder.certificates(certs);
             if (certs == null || certs.length == 0) {
-                XMLSignatureFactory signatureFactory;
-                if (provider == null) {
-                    // Try to install the Santuario Provider - fall back to the JDK provider if this does
-                    // not work
-                    try {
-                        signatureFactory = XMLSignatureFactory.getInstance("DOM", "ApacheXMLDSig");
-                    } catch (NoSuchProviderException ex) {
-                        signatureFactory = XMLSignatureFactory.getInstance("DOM");
-                    }
-                } else {
-                    signatureFactory = XMLSignatureFactory.getInstance("DOM", provider);
-                }
-
-                PublicKey publicKey = X509Util.parseKeyValue((Element) keyValueElement.getParentNode(),
-                        signatureFactory);
+                PublicKey publicKey = getPublicKeyFromKeyValue(keyValueElement);
```

Comment:
What's the reason here not to use XMLSignatureFactory as before?

Reply:
But the keyInfoFactory is used to actually do the unmarshalling in https://github.com/apache/ws-wss4j/blob/21a82b005fe70553698e295c310a660497375bbf/ws-security-dom/src/main/java/org/apache/wss4j/dom/util/X509Util.java#L152
and one of the existing unit tests in EncryptionTest hits this code and finds the PublicKey successfully?

Reply:
OK I realise the problem here is the parent element is not a KeyInfo but a RecipientKeyInfo and the unmarshalling fails